Feature/upload

egrace479 opened the pull request:
Added upload option to Dashboard (csv or xls).
- Updated functions: 
  - `get_filenames`: returns filenames and paths, also patches "Any" subspecies glitch (if user doesn't remove `Any` option before selecting specific subspecies and pressing "Display Images", it will just return images of the selected subspecies).
  -  `test_get_filenames` and `get_images` updated to interact with updated `get_filenames`.
  - `get_data`: also takes boolean indicating if latitude & longitude values are in data to do `lat-lon` manipulations and sets `locality` column with `lat-lon` values if it's not already a column in the dataframe. 
  - `test_get_data`: now checks two variations, one with lat & long, one without.
- Also removed additional taxa info from cat_list (in our test data it was just subspecies data repeated).
- Updated README to reflect upload requirements and added two datasets for testing/demo.
Resolves Issue #23

egrace479 commented:
This is currently what it looks like before upload, then below the line gets replaced by the dashboard as it has been. Clicking on "Upload Data" again allows for a new upload.

![Main Page](https://github.com/Imageomics/dashboard-prototype/assets/38985481/6b1d0916-4e1e-4bbc-9434-a4e89b04accd)

Should I add the column requirements underneath as well? It currently will throw an error (missing column, wrong file type, etc.) and point the user to [how-it-works](https://github.com/Imageomics/dashboard-prototype#how-it-works):
![Error Message (missing column)](https://github.com/Imageomics/dashboard-prototype/assets/38985481/d96a0202-8bde-40b3-9e51-d4fc28b3850b)

hlapp commented on README.md:
```diff
@@ -1,9 +1,24 @@
 # Dashboard Prototype
-Prototype data dashboard using the [Cuthill Gold Standard Dataset](https://datacommons.tdai.osu.edu/dataset.xhtml?persistentId=doi:10.5072/FK2/GZYWNV&version=DRAFT), which was processed from Cuthill, et. al. (original dataset available at [doi:10.5061/dryad.2hp1978](https://doi.org/10.5061/dryad.2hp1978)).
+Prototype data dashboard using the [Cuthill Gold Standard Dataset](https://huggingface.co/datasets/imageomics/Curated_GoldStandard_Hoyal_Cuthill), which was processed from Cuthill, et. al. (original dataset available at [doi:10.5061/dryad.2hp1978](https://doi.org/10.5061/dryad.2hp1978)). Test datasets (the processed version of Cuthill's data with and without filepath URLs) are available in [test_ ...
```
```suggestion
Prototype data dashboard using the [Cuthill Gold Standard Dataset](https://huggingface.co/datasets/imageomics/Curated_GoldStandard_Hoyal_Cuthill), which was processed from Cuthill, et. al. (original dataset available at [doi:10.5061/dryad.2hp1978](https://doi.org/10.5061/dryad.2hp1978)). Test datasets (the processed version of Cuthill's data with and without filepath URLs) are available in [test_data](./test_data).
```

It's probably better not to tie this to a particular branch, and instead just use the relative path.

hlapp commented:
> This is currently what it looks like before upload, then below the line gets replaced by the dashboard as it has been. [...]
> Should I add the column requirements underneath as well? It currently will throw an error (missing column, wrong file type, etc.) and point the user to [how-it-works](https://github.com/Imageomics/dashboard-prototype#how-it-works)

It seems helpful to provide a link to "How it works" before someone runs into an error unintentionally because they didn't fully understand which columns they'd have to have.

hlapp reviewed (commented):
It seems this change set lumps together changes for multiple separate and not necessarily interrelated issues. Generally I'd suggest breaking changes up into meaningful and coherent pieces, this makes review and feedback much easier.

Overall this looks fine but I notice there's a fair amount of code duplication. In the conditional for whether geo-coordinates are available or not, it seems only a small part differs for the two cases, and the rest is duplicated. There are also strings or string prefixes (such as `https://github.com/Imageomics/dashboard-prototype/raw/main/test_data/images/`) that occur many times over, and could just be defined at the beginning and then reused.

egrace479 commented:
> It seems this change set lumps together changes for multiple separate and not necessarily interrelated issues. Generally I'd suggest breaking changes up into meaningful and coherent pieces, this makes review and feedback much easier.
> 
> Overall this looks fine but I notice there's a fair amount of code duplication. In the conditional for whether geo-coordinates are available or not, it seems only a small part differs for the two cases, and the rest is duplicated. There are also strings or string prefixes (such as `https://github.com/Imageomics/dashboard-prototype/raw/main/test_data/images/`) that occur many times over, and could just be defined at the beginning and then reused.

For this PR, would it have been better to do the file retrieval generalization with a revised preloaded dataset and test as its own PR, then move the main div out separately and add the upload feature?

To your second point, I can adjust the geo-coordinate if statement to append the final div to `hist_div` dependent on the boolean. For the string duplication, are you referring to `test_query`? I can adjust that as well.

egrace479 commented:
Updated as per @hlapp's suggestions.

New landing page:
![landing page](https://github.com/Imageomics/dashboard-prototype/assets/38985481/b4b486aa-a276-41e0-8832-f2c4beb16ef6)
Updated Error message page with embedded link that opens in new tab too:
![error page](https://github.com/Imageomics/dashboard-prototype/assets/38985481/19d8b9a1-f1eb-4eb2-bc8b-e6e818d5e402)

johnbradley commented on dashboard.py:
```diff
@@ -83,7 +90,7 @@ def parse_contents(contents, filename):
     processed_df, cat_list = get_data(df, mapping)
     all_species = get_species_options(processed_df)
-    
+    print(mapping)
```
Did you mean to leave this print in?

egrace479 replied:
No, I thought I had removed all of those. Thanks for catching that.

johnbradley reviewed (commented):
Is there a CSV I can test these changes with?

egrace479 commented:
> Is there a CSV I can test these changes with?

In `test_data`: 
- `HCGSD_base_filepath_noMap.csv`: has no lat/lon columns
- `HCGSD_base_filepath.csv`: has base filepath in `file_url` (there's also one with the full filepath, including filename)
- `Hoyal_Cuthill_GoldStandard_metadata_cleaned.csv`: doesn't have `file_url`

hlapp reviewed (approved):
LGTM! 👍

johnbradley commented on dashboard.py:
```diff
@@ -1,147 +1,149 @@
+                            ),
+                dcc.Store(id = 'memory'),
+                html.Hr(),
+                # Set up memory store, will revert on page refresh
+                #dcc.Store(id = 'memory'),
+                #html.Button('Process Data', 
+                #            id = 'memory-btn'),
```
I'm wondering if this commented out code was intentionally left in.

johnbradley commented on dashboard.py:
```diff
@@ -1,147 +1,149 @@
+    if contents is None:
+        raise PreventUpdate
+    content_type, content_string = contents.split(',')
+    #content_string = contents
```
Another stray commented out line that might not be intensional.

johnbradley reviewed (approved):
LGTM. Just a couple commented out bits of code to consider removing.

thompsonmj commented on components/query.py:
```diff
@@ -87,19 +93,20 @@ def get_images(df, subspecies, view, sex, hybrid, num_images):
-    filenames = get_filenames(df, subspecies, view, sex, hybrid, num_images)
+    filenames, filepaths = get_filenames(df, subspecies, view, sex, hybrid, num_images)
     if filenames == 0:
         #print("No Such Images. Please make another selection.")
```
Could delete this one too.
